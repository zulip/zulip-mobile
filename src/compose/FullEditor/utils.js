/* @flow */
import regexValidator from './webLinkValidator';

export const replaceBetween = (text: string, selection: Object, what: string) =>
  text.substring(0, selection.start) + what + text.substring(selection.end);

export const isStringWebLink = (text: string): boolean => {
  const pattern = regexValidator;
  return pattern.test(text);
};
