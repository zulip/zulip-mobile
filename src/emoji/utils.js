/* @flow */
import type { RealmEmojiType, ZulipExtraEmojiType } from '../types';
import { codePointMap } from './emojiMap';

export const getRealmEmojiWithName = (realmEmoji: Object, emojiName: string): RealmEmojiType =>
  realmEmoji[Object.keys(realmEmoji).find(key => realmEmoji[key].name === emojiName)];

export const getZulipExtraEmojiWithName = (
  zulipExtraEmojis: Object,
  emojiName: string,
): ZulipExtraEmojiType => zulipExtraEmojis[emojiName];

export const getEmojiTypeFromName = (
  realmEmoji: Object,
  zulipExtraEmojis: Object,
  emojiName: string,
): string | null => {
  if (getZulipExtraEmojiWithName(zulipExtraEmojis, emojiName)) {
    return 'zulip_extra_emoji';
  } else if (getRealmEmojiWithName(realmEmoji, emojiName)) {
    return 'realm_emoji';
  } else if (codePointMap[emojiName]) {
    return 'unicode_emoji';
  }
  return null;
};

export const getEmojiCodeFromName = (
  realmEmoji: Object,
  zulipExtraEmojis: Object,
  emojiName: string,
): string | null => {
  if (getZulipExtraEmojiWithName(zulipExtraEmojis, emojiName)) {
    return emojiName;
  }
  const emoji = getRealmEmojiWithName(realmEmoji, emojiName);
  if (emoji) {
    return emoji.id.toString();
  } else if (codePointMap[emojiName]) {
    return codePointMap[emojiName];
  }

  return null;
};
