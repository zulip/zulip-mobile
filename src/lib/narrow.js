export const homeNarrow = {};

export const mentionedNarrow = [{
  negated: false,
  operator: 'is',
  operand: 'mentioned',
}];

export const privateNarrow = (emails) => {
  if (!emails) {
    return [{
      negated: false,
      operator: 'is',
      operand: 'private',
    }];
  }
  return [{
    negated: false,
    operator: 'pm-with',
    operand: emails.join(),
  }];
};

export const starredNarrow = [{
  negated: false,
  operator: 'is',
  operand: 'starred',
}];

export const streamNarrow = (stream) => [{
  negated: false,
  operator: 'stream',
  operand: stream,
}];

export const topicNarrow = (stream, topic) => [
  {
    negated: false,
    operator: 'stream',
    operand: stream,
  },
  {
    negated: false,
    operator: 'topic',
    operand: topic,
  },
];
