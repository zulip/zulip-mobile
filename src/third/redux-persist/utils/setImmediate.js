const hasNativeSupport = typeof global !== 'undefined' && typeof global.setImmediate !== 'undefined'
const setImmediate = hasNativeSupport ? (fn, ms) => global.setImmediate(fn, ms) : (fn, ms) => setTimeout(fn, ms)

export default setImmediate
