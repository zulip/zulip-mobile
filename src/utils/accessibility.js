/* @flow */
export const textWithUnreadCount = (text: string, unreadCount: number = 0): string => {
  if (unreadCount === 0) {
    return text;
  }

  return `${text}, ${unreadCount} unread message${unreadCount > 1 ? 's' : ''}`;
};
