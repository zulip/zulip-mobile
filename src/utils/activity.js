/* @flow */
import differenceInSeconds from 'date-fns/difference_in_seconds';

import type { Auth } from '../types';
import { focusPing } from '../api';

const APP_ACTIVITY_UPDATE_FREQUENCY = 60;
const USER_INPUT_ACTIVITY_UPDATE_FREQUENCY = 5;

let lastAppActivity = new Date();
let lastInputActivity = new Date();

export const registerAppActivity = (auth: Auth, hasFocus: boolean = true) => {
  if (differenceInSeconds(new Date(), lastAppActivity) > APP_ACTIVITY_UPDATE_FREQUENCY) {
    focusPing(auth, hasFocus, false);
    lastAppActivity = new Date();
  }
};

export const registerUserInputActivity = (auth: Auth) => {
  if (differenceInSeconds(new Date(), lastInputActivity) > USER_INPUT_ACTIVITY_UPDATE_FREQUENCY) {
    focusPing(auth, true, true);
    lastInputActivity = new Date();
  }
};
