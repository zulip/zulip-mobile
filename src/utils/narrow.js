/* @flow strict-local */
import isEqual from 'lodash.isequal';
import unescape from 'lodash.unescape';

import type { Narrow, Message, Outbox } from '../types';
import { normalizeRecipients } from './recipient';

export const isSameNarrow = (narrow1: Narrow, narrow2: Narrow): boolean =>
  Array.isArray(narrow1) && Array.isArray(narrow2) && isEqual(narrow1, narrow2);

export const parseNarrowString = (narrowStr: string): Narrow => JSON.parse(unescape(narrowStr));

export const HOME_NARROW: Narrow = [];

export const HOME_NARROW_STR: string = '[]';

export const privateNarrow = (email: string): Narrow => [
  {
    operator: 'pm-with',
    operand: email,
  },
];

export const groupNarrow = (emails: string[]): Narrow => [
  {
    operator: 'pm-with',
    operand: emails.join(),
  },
];

export const specialNarrow = (operand: string): Narrow => [
  {
    operator: 'is',
    operand,
  },
];

export const STARRED_NARROW = specialNarrow('starred');

export const STARRED_NARROW_STR = JSON.stringify(STARRED_NARROW);

export const MENTIONED_NARROW = specialNarrow('mentioned');

export const MENTIONED_NARROW_STR = JSON.stringify(MENTIONED_NARROW);

export const ALL_PRIVATE_NARROW = specialNarrow('private');

export const ALL_PRIVATE_NARROW_STR = JSON.stringify(ALL_PRIVATE_NARROW);

export const streamNarrow = (stream: string): Narrow => [
  {
    operator: 'stream',
    operand: stream,
  },
];

export const topicNarrow = (stream: string, topic: string): Narrow => [
  {
    operator: 'stream',
    operand: stream,
  },
  {
    operator: 'topic',
    operand: topic,
  },
];

export const SEARCH_NARROW = (query: string): Narrow => [
  {
    operator: 'search',
    operand: query,
  },
];

type NarrowCases<T> = {|
  home: () => T,
  pm: (email: string) => T,
  groupPm: (emails: string[]) => T,
  starred: () => T,
  mentioned: () => T,
  allPrivate: () => T,
  stream: (name: string) => T,
  topic: (streamName: string, topic: string) => T,
  search: (query: string) => T,
|};

/* prettier-ignore */
export function caseNarrow<T>(narrow: Narrow, cases: NarrowCases<T>): T {
  const err = (): empty => {
    throw new Error(`bad narrow: ${JSON.stringify(narrow)}`);
  };

  switch (narrow.length) {
    case 0: return cases.home();
    case 1:
      switch (narrow[0].operator) {
        case 'pm-with':
          if (narrow[0].operand.indexOf(',') < 0) {
            return cases.pm(narrow[0].operand);
          } else { /* eslint-disable-line */
            const emails = narrow[0].operand.split(',');
            return cases.groupPm(emails);
          }
        case 'is':
          switch (narrow[0].operand) {
            case 'starred': return cases.starred();
            case 'mentioned': return cases.mentioned();
            case 'private': return cases.allPrivate();
            default: return err();
          }
        case 'stream': return cases.stream(narrow[0].operand);
        case 'search': return cases.search(narrow[0].operand);
        default: return err();
      }
    case 2: return cases.topic(narrow[0].operand, narrow[1].operand);
    default: return err();
  }
}

export function caseNarrowPartial<T>(narrow: Narrow, cases: $Shape<NarrowCases<T>>): T {
  const err = (type: string) => (): empty => {
    throw new Error(`unexpected ${type} narrow: ${JSON.stringify(narrow)}`);
  };
  return caseNarrow(
    narrow,
    Object.assign(
      ({
        home: err('home'),
        pm: err('PM'),
        groupPm: err('group PM'),
        starred: err('starred'),
        mentioned: err('mentions'),
        allPrivate: err('all-private'),
        stream: err('stream'),
        topic: err('topic'),
        search: err('search'),
      }: NarrowCases<T>),
      cases,
    ),
  );
}

export function caseNarrowDefault<T>(
  narrow: Narrow,
  cases: $Shape<NarrowCases<T>>,
  defaultCase: () => T,
): T {
  return caseNarrow(
    narrow,
    Object.assign(
      ({
        home: defaultCase,
        pm: defaultCase,
        groupPm: defaultCase,
        starred: defaultCase,
        mentioned: defaultCase,
        allPrivate: defaultCase,
        stream: defaultCase,
        topic: defaultCase,
        search: defaultCase,
      }: NarrowCases<T>),
      cases,
    ),
  );
}

export const isHomeNarrow = (narrow?: Narrow): boolean =>
  !!narrow && caseNarrowDefault(narrow, { home: () => true }, () => false);

export const isPrivateNarrow = (narrow?: Narrow): boolean =>
  !!narrow && caseNarrowDefault(narrow, { pm: () => true }, () => false);

export const isGroupNarrow = (narrow?: Narrow): boolean =>
  !!narrow && caseNarrowDefault(narrow, { groupPm: () => true }, () => false);

/**
 * The recipients' emails if a group PM narrow; else error.
 *
 * Any use of this probably means something higher up should be refactored
 * to use caseNarrow.
 */
export const emailsOfGroupNarrow = (narrow: Narrow): string[] =>
  caseNarrowPartial(narrow, { groupPm: emails => emails });

/**
 * The recipient's email if a Private narrow; else error.
 */
export const emailOfPrivateNarrow = (narrow: Narrow): string =>
  caseNarrowPartial(narrow, { pm: email => email });

export const isPrivateOrGroupNarrow = (narrow?: Narrow): boolean =>
  !!narrow && caseNarrowDefault(narrow, { pm: () => true, groupPm: () => true }, () => false);

export const isSpecialNarrow = (narrow?: Narrow): boolean =>
  !!narrow
  && caseNarrowDefault(
    narrow,
    { starred: () => true, mentioned: () => true, allPrivate: () => true },
    () => false,
  );

export const isAllPrivateNarrow = (narrow?: Narrow): boolean =>
  !!narrow && caseNarrowDefault(narrow, { allPrivate: () => true }, () => false);

export const isStreamNarrow = (narrow?: Narrow): boolean =>
  !!narrow && caseNarrowDefault(narrow, { stream: () => true }, () => false);

export const isTopicNarrow = (narrow?: Narrow): boolean =>
  !!narrow && caseNarrowDefault(narrow, { topic: () => true }, () => false);

export const isStreamOrTopicNarrow = (narrow?: Narrow): boolean =>
  !!narrow && caseNarrowDefault(narrow, { stream: () => true, topic: () => true }, () => false);

export const isSearchNarrow = (narrow?: Narrow): boolean =>
  !!narrow && caseNarrowDefault(narrow, { search: () => true }, () => false);

/** (For search narrows, just returns false.) */
export const isMessageInNarrow = (message: Message, narrow: Narrow, ownEmail: string): boolean => {
  const matchRecipients = (emails: string[]) => {
    const normalizedRecipients = normalizeRecipients(message.display_recipient);
    const normalizedNarrow = [...emails, ownEmail].sort().join(',');
    return normalizedRecipients === ownEmail || normalizedRecipients === normalizedNarrow;
  };

  const { flags } = message;
  if (!flags) {
    throw new Error('`message.flags` should be defined.');
  }

  return caseNarrow(narrow, {
    home: () => true,
    stream: name => name === message.display_recipient,
    topic: (streamName, topic) =>
      streamName === message.display_recipient && topic === message.subject,
    pm: email => matchRecipients([email]),
    groupPm: matchRecipients,
    starred: () => flags.includes('starred'),
    mentioned: () => flags.includes('mentioned') || flags.includes('wildcard_mentioned'),
    allPrivate: () => message.type === 'private',
    search: () => false,
  });
};

export const canSendToNarrow = (narrow: Narrow): boolean =>
  caseNarrow(narrow, {
    pm: () => true,
    groupPm: () => true,
    stream: () => true,
    topic: () => true,
    home: () => false,
    starred: () => false,
    mentioned: () => false,
    allPrivate: () => false,
    search: () => false,
  });

/** True just if `haystack` contains all possible messages in `needle`. */
export const narrowContains = (haystack: Narrow, needle: Narrow): boolean => {
  if (isHomeNarrow(haystack)) {
    return true;
  }
  if (isAllPrivateNarrow(haystack) && isPrivateOrGroupNarrow(needle)) {
    return true;
  }
  if (isStreamNarrow(haystack) && needle[0].operand === haystack[0].operand) {
    return true;
  }
  return JSON.stringify(needle) === JSON.stringify(haystack);
};

export const getNarrowFromMessage = (message: Message | Outbox, ownEmail: string) => {
  if (Array.isArray(message.display_recipient)) {
    const recipient =
      message.display_recipient.length > 1
        ? message.display_recipient.filter(x => x.email !== ownEmail)
        : message.display_recipient;
    return groupNarrow(recipient.map(x => x.email));
  }

  if (message.subject && message.subject.length) {
    return topicNarrow(message.display_recipient, message.subject);
  }

  return streamNarrow(message.display_recipient);
};
