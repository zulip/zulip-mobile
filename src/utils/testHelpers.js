/* @flow strict-local */
import type { Narrow } from '../types';

export const navStateWithNarrow = (narrow: Narrow): {| nav: mixed |} => ({
  nav: {
    index: 0,
    routes: [
      {
        routeName: 'chat',
        params: {
          narrow,
        },
      },
    ],
  },
});

export const defaultNav = {
  index: 0,
  routes: [{ routeName: 'chat' }],
};

export const otherNav = {
  index: 1,
  routes: [{ routeName: 'main' }, { routeName: 'account' }],
};
