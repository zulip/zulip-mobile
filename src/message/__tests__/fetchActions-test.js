import mockStore from 'redux-mock-store'; // eslint-disable-line

import {
  backgroundFetchMessages,
  fetchMessages,
  fetchMessagesAtFirstUnread,
  fetchOlder,
  fetchNewer,
} from '../fetchActions';
import { streamNarrow, homeNarrow, homeNarrowStr } from '../../utils/narrow';
import { navStateWithNarrow } from '../../utils/testHelpers';

const narrow = streamNarrow('some stream');
const streamNarrowStr = JSON.stringify(narrow);

global.FormData = class FormData {};

describe('fetchActions', () => {
  afterEach(() => {
    fetch.reset();
  });

  describe('backgroundFetchMessages', () => {
    test('message fetch success action is dispatched after successful fetch', async () => {
      const store = mockStore({
        ...navStateWithNarrow(homeNarrow),
        accounts: [
          {
            realm: 'https://example.com',
          },
        ],
        messages: {
          [streamNarrowStr]: [{ id: 1 }],
        },
      });
      const response = { messages: [{ id: 1 }, { id: 2 }, { id: 3 }], result: 'success' };
      fetch.mockResponseSuccess(JSON.stringify(response));

      await store.dispatch(backgroundFetchMessages(homeNarrow, 0, 1, 1, true));
      const actions = store.getActions();

      expect(actions.length).toBe(1);
      expect(actions[0].type).toBe('MESSAGE_FETCH_COMPLETE');
    });
  });

  describe('fetchMessages', () => {
    test('when messages to be fetched both before and after anchor, fetchingOlder and fetchingNewer is true', () => {
      const store = mockStore({
        ...navStateWithNarrow(homeNarrow),
        messages: {
          [streamNarrowStr]: [{ id: 1 }],
        },
      });

      store.dispatch(fetchMessages(homeNarrow, 0, 1, 1, true));
      const actions = store.getActions();

      expect(actions.length).toBe(1);
      expect(actions[0].type).toBe('MESSAGE_FETCH_START');
    });

    test('when no messages to be fetched before the anchor, fetchingOlder is false', () => {
      const store = mockStore({
        ...navStateWithNarrow(homeNarrow),
        messages: {
          [streamNarrowStr]: [{ id: 1 }],
        },
      });

      store.dispatch(fetchMessages(homeNarrow, 0, -1, 1, true));
      const actions = store.getActions();

      expect(actions.length).toBe(1);
      expect(actions[0].type).toBe('MESSAGE_FETCH_START');
    });

    test('when no messages to be fetched after the anchor, fetchingNewer is false', () => {
      const store = mockStore({
        ...navStateWithNarrow(homeNarrow),
        messages: {
          [streamNarrowStr]: [{ id: 1 }],
        },
      });

      store.dispatch(fetchMessages(homeNarrow, 0, 1, -1, true));
      const actions = store.getActions();

      expect(actions.length).toBe(1);
      expect(actions[0].type).toBe('MESSAGE_FETCH_START');
    });
  });

  describe('fetchMessagesAtFirstUnread', () => {
    test('message fetch start action is dispatched with fetchingOlder and fetchingNewer true', () => {
      const store = mockStore({
        ...navStateWithNarrow(homeNarrow),
        messages: {
          [streamNarrowStr]: [{ id: 1 }],
        },
      });

      store.dispatch(fetchMessagesAtFirstUnread(homeNarrow));
      const actions = store.getActions();

      expect(actions.length).toBe(1);
      expect(actions[0].type).toBe('MESSAGE_FETCH_START');
    });
  });

  describe('fetchOlder', () => {
    test('message fetch start action is dispatched with fetchingOlder true', () => {
      const store = mockStore({
        session: {
          needsInitialFetch: false,
        },
        caughtUp: {
          [homeNarrowStr]: { older: false },
        },
        ...navStateWithNarrow(homeNarrow),
        messages: {
          [streamNarrowStr]: [{ id: 2 }],
          [homeNarrowStr]: [{ id: 1 }, { id: 2 }],
        },
        fetching: {
          [homeNarrowStr]: { older: false, newer: false },
        },
      });

      store.dispatch(fetchOlder(homeNarrow));
      const actions = store.getActions();

      expect(actions.length).toBe(1);
      expect(actions[0].type).toBe('MESSAGE_FETCH_START');
    });

    test('when caughtUp older is true, no action is dispatched', () => {
      const store = mockStore({
        session: {
          needsInitialFetch: false,
        },
        caughtUp: {
          [homeNarrowStr]: { older: true },
        },
        ...navStateWithNarrow(homeNarrow),
        messages: {
          [streamNarrowStr]: [{ id: 2 }],
          [homeNarrowStr]: [{ id: 1 }, { id: 2 }],
        },
        fetching: {
          [homeNarrowStr]: { older: false, newer: false },
        },
      });

      store.dispatch(fetchOlder(homeNarrow));
      const actions = store.getActions();

      expect(actions.length).toBe(0);
    });

    test('when fetchingOlder older is true, no action is dispatched', () => {
      const store = mockStore({
        session: {
          needsInitialFetch: false,
        },
        caughtUp: {
          [homeNarrowStr]: { older: false },
        },
        ...navStateWithNarrow(homeNarrow),
        messages: {
          [streamNarrowStr]: [{ id: 2 }],
          [homeNarrowStr]: [{ id: 1 }, { id: 2 }],
        },
        fetching: {
          [homeNarrowStr]: { older: true, newer: false },
        },
      });

      store.dispatch(fetchOlder(homeNarrow));
      const actions = store.getActions();

      expect(actions.length).toBe(0);
    });

    test('when needsInitialFetch is true, no action is dispatched', () => {
      const store = mockStore({
        session: {
          needsInitialFetch: true,
        },
        caughtUp: {
          [homeNarrowStr]: { older: false },
        },
        ...navStateWithNarrow(homeNarrow),
        messages: {
          [streamNarrowStr]: [{ id: 2 }],
          [homeNarrowStr]: [{ id: 1 }, { id: 2 }],
        },
        fetching: {},
      });

      store.dispatch(fetchOlder(homeNarrow));
      const actions = store.getActions();

      expect(actions.length).toBe(0);
    });
  });

  describe('fetchNewer', () => {
    test('message fetch start action is dispatched with fetchingNewer true', () => {
      const store = mockStore({
        session: {
          needsInitialFetch: false,
        },
        caughtUp: {
          [homeNarrowStr]: { newer: false },
        },
        ...navStateWithNarrow(homeNarrow),
        messages: {
          [streamNarrowStr]: [{ id: 2 }],
          [homeNarrowStr]: [{ id: 1 }, { id: 2 }],
        },
        fetching: {},
      });

      store.dispatch(fetchNewer(homeNarrow));
      const actions = store.getActions();

      expect(actions.length).toBe(1);
      expect(actions[0].type).toBe('MESSAGE_FETCH_START');
    });

    test('when caughtUp newer is true, no action is dispatched', () => {
      const store = mockStore({
        session: {
          needsInitialFetch: false,
        },
        caughtUp: {
          [homeNarrowStr]: { newer: true },
        },
        ...navStateWithNarrow(homeNarrow),
        messages: {
          [streamNarrowStr]: [{ id: 2 }],
          [homeNarrowStr]: [{ id: 1 }, { id: 2 }],
        },
        fetching: {},
      });

      store.dispatch(fetchNewer(homeNarrow));
      const actions = store.getActions();

      expect(actions.length).toBe(0);
    });

    test('when fetching.newer is true, no action is dispatched', () => {
      const store = mockStore({
        session: {
          needsInitialFetch: false,
        },
        caughtUp: {
          [homeNarrowStr]: { newer: false },
        },
        ...navStateWithNarrow(homeNarrow),
        messages: {
          [streamNarrowStr]: [{ id: 2 }],
          [homeNarrowStr]: [{ id: 1 }, { id: 2 }],
        },
        fetching: {
          homeNarrow: { older: false, newer: true },
        },
      });

      store.dispatch(fetchNewer(homeNarrow));
      const actions = store.getActions();

      expect(actions.length).toBe(1);
      expect(actions[0].type).toBe('MESSAGE_FETCH_START');
    });

    test('when needsInitialFetch is true, no action is dispatched', () => {
      const store = mockStore({
        session: {
          needsInitialFetch: true,
        },
        caughtUp: {
          [homeNarrowStr]: { newer: false },
        },
        fetching: {},
        ...navStateWithNarrow(homeNarrow),
        messages: {
          [streamNarrowStr]: [{ id: 2 }],
          [homeNarrowStr]: [{ id: 1 }, { id: 2 }],
        },
      });

      store.dispatch(fetchNewer(homeNarrow));
      const actions = store.getActions();

      expect(actions.length).toBe(0);
    });
  });
});
