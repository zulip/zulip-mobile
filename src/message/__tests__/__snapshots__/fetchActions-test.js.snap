// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchActions backgroundFetchMessages message fetch success action is dispatched after successful fetch 1`] = `
Array [
  Object {
    "anchor": 1,
    "messages": Array [
      Object {
        "id": 1,
      },
      Object {
        "id": 2,
      },
      Object {
        "id": 3,
      },
    ],
    "narrow": Array [],
    "numAfter": 1,
    "numBefore": 1,
    "replaceExisting": false,
    "type": "MESSAGE_FETCH_COMPLETE",
  },
]
`;

exports[`fetchActions fetchMessages when messages to be fetched both before and after anchor, fetchingOlder and fetchingNewer is true 1`] = `
Array [
  Object {
    "narrow": Array [],
    "numAfter": 1,
    "numBefore": 1,
    "type": "MESSAGE_FETCH_START",
  },
]
`;

exports[`fetchActions fetchMessages when no messages to be fetched after the anchor, fetchingNewer is false 1`] = `
Array [
  Object {
    "narrow": Array [],
    "numAfter": -1,
    "numBefore": 1,
    "type": "MESSAGE_FETCH_START",
  },
]
`;

exports[`fetchActions fetchMessages when no messages to be fetched before the anchor, fetchingOlder is false 1`] = `
Array [
  Object {
    "narrow": Array [],
    "numAfter": 1,
    "numBefore": -1,
    "type": "MESSAGE_FETCH_START",
  },
]
`;

exports[`fetchActions fetchMessagesAtFirstUnread message fetch start action is dispatched with fetchingOlder and fetchingNewer true 1`] = `
Array [
  Object {
    "narrow": Array [],
    "numAfter": 20,
    "numBefore": 20,
    "type": "MESSAGE_FETCH_START",
  },
]
`;

exports[`fetchActions fetchNewer message fetch start action is dispatched with fetchingNewer true 1`] = `
Array [
  Object {
    "narrow": Array [],
    "numAfter": 20,
    "numBefore": 0,
    "type": "MESSAGE_FETCH_START",
  },
]
`;

exports[`fetchActions fetchNewer when caughtUp newer is true, no action is dispatched 1`] = `Array []`;

exports[`fetchActions fetchNewer when fetching.newer is true, no action is dispatched 1`] = `
Array [
  Object {
    "narrow": Array [],
    "numAfter": 20,
    "numBefore": 0,
    "type": "MESSAGE_FETCH_START",
  },
]
`;

exports[`fetchActions fetchNewer when needsInitialFetch is true, no action is dispatched 1`] = `Array []`;

exports[`fetchActions fetchOlder message fetch start action is dispatched with fetchingOlder true 1`] = `
Array [
  Object {
    "narrow": Array [],
    "numAfter": 0,
    "numBefore": 20,
    "type": "MESSAGE_FETCH_START",
  },
]
`;

exports[`fetchActions fetchOlder when caughtUp older is true, no action is dispatched 1`] = `Array []`;

exports[`fetchActions fetchOlder when fetchingOlder older is true, no action is dispatched 1`] = `
Array [
  Object {
    "narrow": Array [],
    "numAfter": 0,
    "numBefore": 20,
    "type": "MESSAGE_FETCH_START",
  },
]
`;

exports[`fetchActions fetchOlder when needsInitialFetch is true, no action is dispatched 1`] = `Array []`;
